From 40d6e8b23eba85f5459c7b53d5eb1f69d1d425b6 Mon Sep 17 00:00:00 2001
From: zhengq <qi.zheng@intel.com>
Date: Tue, 31 Mar 2015 22:42:54 +0800
Subject: [PATCH] Pinctrl: do pad mode switch conditionally on CHT.

For Cherryview platform, the pad mode is set to GPIO mode
in gpio_request, then switch it back to native mode in
gpio_free. It is necessary for some drivers, such as HSU, to
do pinmux switch during resume/suspend. However, this solution
introduces power over consumption issue because not all the pins
used by different platforms need switch it back to native mode
when gpio_free.
Add an item "switch_en" to specify whether this pinmux switch
in gpio_request/free is needed.
Currently, only HSU RXD/RTS pins are enabled for the switch.

Change-Id: I9a9d72943ef62191159638203aed96f1143824dd
Tracked-On: https://jira01.devtools.intel.com/browse/AKP-1334
Signed-off-by: zhengq <qi.zheng@intel.com>
---
 drivers/pinctrl/pinctrl-cherryview.c | 607 ++++++++++++++++++-----------------
 1 file changed, 310 insertions(+), 297 deletions(-)

diff --git a/drivers/pinctrl/pinctrl-cherryview.c b/drivers/pinctrl/pinctrl-cherryview.c
index 57aa1b14154f..b695ef601fd8 100644
--- a/drivers/pinctrl/pinctrl-cherryview.c
+++ b/drivers/pinctrl/pinctrl-cherryview.c
@@ -320,6 +320,11 @@ struct gpio_pad_info {
 	/* Interrupt line selected (0~15), -1 if not interruptible. */
 	int interrupt_line;
 
+	/*function mode switch support. Once enable, gpio_request will force set
+	* the pad mode to GPIO, gpio_free will switch it back to native mode.
+	* It is disabled in default. */
+	int switch_en;
+
 };
 
 struct gpio_bank_pnp {
@@ -335,344 +340,344 @@ struct gpio_bank_pnp {
  */
 static struct gpio_pad_info north_pads_info[] = {
 	/*0~14*/
-	[GPIO_DFX_0]		= { 0,		0,	-1 },
-	[GPIO_DFX_3]		= { 0,		1,	-1 },
-	[GPIO_DFX_7]		= { 0,		2,	-1 },
-	[GPIO_DFX_1]		= { 0,		3,	-1 },
-	[GPIO_DFX_5]		= { 0,		4,	-1 },
-	[GPIO_DFX_4]		= { 0,		5,	-1 },
-	[GPIO_DFX_8]		= { 0,		6,	-1 },
-	[GPIO_DFX_2]		= { 0,		7,	-1 },
-	[GPIO_DFX_6]		= { 0,		8,	-1 },
-	[9]					= { -1,		-1,	-1 },
-	[10]				= { -1,		-1,	-1 },
-	[11]				= { -1,		-1,	-1 },
-	[12]				= { -1,		-1,	-1 },
-	[13]				= { -1,		-1,	-1 },
-	[14]				= { -1,		-1,	-1 },
+	[GPIO_DFX_0]		= { 0,		0,	-1, -1},
+	[GPIO_DFX_3]		= { 0,		1,	-1, -1},
+	[GPIO_DFX_7]		= { 0,		2,	-1, -1},
+	[GPIO_DFX_1]		= { 0,		3,	-1, -1},
+	[GPIO_DFX_5]		= { 0,		4,	-1, -1},
+	[GPIO_DFX_4]		= { 0,		5,	-1, -1},
+	[GPIO_DFX_8]		= { 0,		6,	-1, -1},
+	[GPIO_DFX_2]		= { 0,		7,	-1, -1},
+	[GPIO_DFX_6]		= { 0,		8,	-1, -1},
+	[9]					= { -1,		-1,	-1, -1},
+	[10]				= { -1,		-1,	-1, -1},
+	[11]				= { -1,		-1,	-1, -1},
+	[12]				= { -1,		-1,	-1, -1},
+	[13]				= { -1,		-1,	-1, -1},
+	[14]				= { -1,		-1,	-1, -1},
 
 	/*15~29*/
-	[GPIO_SUS0]			= { 1,		0,	-1 },
-	[SEC_GPIO_SUS10]	= { 1,		1,	-1 },
-	[GPIO_SUS3]			= { 1,		2,	-1 },
-	[GPIO_SUS7]			= { 1,		3,	-1 },
-	[GPIO_SUS1]			= { 1,		4,	-1 },
-	[GPIO_SUS5]			= { 1,		5,	-1 },
-	[SEC_GPIO_SUS11]	= { 1,		6,	-1 },
-	[GPIO_SUS4]			= { 1,		7,	-1 },
-	[SEC_GPIO_SUS8]		= { 1,		8,	-1 },
-	[GPIO_SUS2]			= { 1,		9,	-1 },
-	[GPIO_SUS6]			= { 1,		10,	-1 },
-	[CX_PREQ_B]			= { 1,		11,	-1 },
-	[SEC_GPIO_SUS9]		= { 1,		12,	-1 },
-	[28]				= { -1,		-1,	-1 },
-	[29]				= { -1,		-1,	-1 },
+	[GPIO_SUS0]			= { 1,		0,	-1, -1},
+	[SEC_GPIO_SUS10]	= { 1,		1,	-1, -1},
+	[GPIO_SUS3]			= { 1,		2,	-1, -1},
+	[GPIO_SUS7]			= { 1,		3,	-1, -1},
+	[GPIO_SUS1]			= { 1,		4,	-1, -1},
+	[GPIO_SUS5]			= { 1,		5,	-1, -1},
+	[SEC_GPIO_SUS11]	= { 1,		6,	-1, -1},
+	[GPIO_SUS4]			= { 1,		7,	-1, -1},
+	[SEC_GPIO_SUS8]		= { 1,		8,	-1, -1},
+	[GPIO_SUS2]			= { 1,		9,	-1, -1},
+	[GPIO_SUS6]			= { 1,		10,	-1, -1},
+	[CX_PREQ_B]			= { 1,		11,	-1, -1},
+	[SEC_GPIO_SUS9]		= { 1,		12,	-1, -1},
+	[28]				= { -1,		-1,	-1, -1},
+	[29]				= { -1,		-1,	-1, -1},
 
 	/*30~44*/
-	[TRST_B]			= { 2,		0,	-1 },
-	[TCK]				= { 2,		1,	-1 },
-	[PROCHOT_B]			= { 2,		2,	-1 },
-	[SVIDO_DATA]		= { 2,		3,	-1 },
-	[TMS]				= { 2,		4,	-1 },
-	[CX_PRDY_B_2]		= { 2,		5,	-1 },
-	[TDO_2]				= { 2,		6,	-1 },
-	[CX_PRDY_B]			= { 2,		7,	-1 },
-	[SVIDO_ALERT_B]		= { 2,		8,	-1 },
-	[TDO]				= { 2,		9,	-1 },
-	[SVIDO_CLK]			= { 2,		10,	-1 },
-	[TDI]				= { 2,		11,	-1 },
-	[42]				= { -1,		-1,	-1 },
-	[43]				= { -1,		-1,	-1 },
-	[44]				= { -1,		-1,	-1 },
+	[TRST_B]			= { 2,		0,	-1, -1},
+	[TCK]				= { 2,		1,	-1, -1},
+	[PROCHOT_B]			= { 2,		2,	-1, -1},
+	[SVIDO_DATA]		= { 2,		3,	-1, -1},
+	[TMS]				= { 2,		4,	-1, -1},
+	[CX_PRDY_B_2]		= { 2,		5,	-1, -1},
+	[TDO_2]				= { 2,		6,	-1, -1},
+	[CX_PRDY_B]			= { 2,		7,	-1, -1},
+	[SVIDO_ALERT_B]		= { 2,		8,	-1, -1},
+	[TDO]				= { 2,		9,	-1, -1},
+	[SVIDO_CLK]			= { 2,		10,	-1, -1},
+	[TDI]				= { 2,		11,	-1, -1},
+	[42]				= { -1,		-1,	-1, -1},
+	[43]				= { -1,		-1,	-1, -1},
+	[44]				= { -1,		-1,	-1, -1},
 
 	/*45~59*/
-	[GP_CAMERASB_05]	= { 3,		0,	-1 },
-	[GP_CAMERASB_02]	= { 3,		1,	-1 },
-	[GP_CAMERASB_08]	= { 3,		2,	-1 },
-	[GP_CAMERASB_00]	= { 3,		3,	-1 },
-	[GP_CAMERASB_06]	= { 3,		4,	-1 },
-	[GP_CAMERASB_10]	= { 3,		5,	-1 },
-	[GP_CAMERASB_03]	= { 3,		6,	-1 },
-	[GP_CAMERASB_09]	= { 3,		7,	-1 },
-	[GP_CAMERASB_01]	= { 3,		8,	-1 },
-	[GP_CAMERASB_07]	= { 3,		9,	-1 },
-	[GP_CAMERASB_11]	= { 3,		10,	-1 },
-	[GP_CAMERASB_04]	= { 3,		11,	-1 },
-	[57]				= { -1,		-1,	-1 },
-	[58]				= { -1,		-1,	-1 },
-	[59]				= { -1,		-1,	-1 },
+	[GP_CAMERASB_05]	= { 3,		0,	-1, -1},
+	[GP_CAMERASB_02]	= { 3,		1,	-1, -1},
+	[GP_CAMERASB_08]	= { 3,		2,	-1, -1},
+	[GP_CAMERASB_00]	= { 3,		3,	-1, -1},
+	[GP_CAMERASB_06]	= { 3,		4,	-1, -1},
+	[GP_CAMERASB_10]	= { 3,		5,	-1, -1},
+	[GP_CAMERASB_03]	= { 3,		6,	-1, -1},
+	[GP_CAMERASB_09]	= { 3,		7,	-1, -1},
+	[GP_CAMERASB_01]	= { 3,		8,	-1, -1},
+	[GP_CAMERASB_07]	= { 3,		9,	-1, -1},
+	[GP_CAMERASB_11]	= { 3,		10,	-1, -1},
+	[GP_CAMERASB_04]	= { 3,		11,	-1, -1},
+	[57]				= { -1,		-1,	-1, -1},
+	[58]				= { -1,		-1,	-1, -1},
+	[59]				= { -1,		-1,	-1, -1},
 
 	/*60~72*/
-	[PANEL0_BKLTEN]		= { 4,		0,	-1 },
-	[HV_DDI0_HPD]		= { 4,		1,	-1 },
-	[HV_DDI2_DDC_SDA]	= { 4,		2,	-1 },
-	[PANEL1_BKLTCTL]	= { 4,		3,	-1 },
-	[HV_DDI1_HPD]		= { 4,		4,	-1 },
-	[PANEL0_BKLTCTL]	= { 4,		5,	-1 },
-	[HV_DDI0_DDC_SDA]	= { 4,		6,	-1 },
-	[HV_DDI2_DDC_SCL]	= { 4,		7,	-1 },
-	[HV_DDI2_HPD]		= { 4,		8,	-1 },
-	[PANEL1_VDDEN]		= { 4,		9,	-1 },
-	[PANEL1_BKLTEN]		= { 4,		10,	-1 },
-	[HV_DDI0_DDC_SCL]	= { 4,		11,	-1 },
-	[PANEL0_VDDEN]		= { 4,		12,	-1 },
+	[PANEL0_BKLTEN]		= { 4,		0,	-1, -1},
+	[HV_DDI0_HPD]		= { 4,		1,	-1, -1},
+	[HV_DDI2_DDC_SDA]	= { 4,		2,	-1, -1},
+	[PANEL1_BKLTCTL]	= { 4,		3,	-1, -1},
+	[HV_DDI1_HPD]		= { 4,		4,	-1, -1},
+	[PANEL0_BKLTCTL]	= { 4,		5,	-1, -1},
+	[HV_DDI0_DDC_SDA]	= { 4,		6,	-1, -1},
+	[HV_DDI2_DDC_SCL]	= { 4,		7,	-1, -1},
+	[HV_DDI2_HPD]		= { 4,		8,	-1, -1},
+	[PANEL1_VDDEN]		= { 4,		9,	-1, -1},
+	[PANEL1_BKLTEN]		= { 4,		10,	-1, -1},
+	[HV_DDI0_DDC_SCL]	= { 4,		11,	-1, -1},
+	[PANEL0_VDDEN]		= { 4,		12,	-1, -1},
 };
 
 static struct gpio_pad_info southeast_pads_info[] = {
 	/*0~14*/
-	[MF_PLT_CLK0]		= { 0,		0,	-1 },
-	[PWM1]				= { 0,		1,	-1 },
-	[MF_PLT_CLK1]		= { 0,		2,	-1 },
-	[MF_PLT_CLK4]		= { 0,		3,	-1 },
-	[MF_PLT_CLK3]		= { 0,		4,	-1 },
-	[PWM0]				= { 0,		5,	-1 },
-	[MF_PLT_CLK5]		= { 0,		6,	-1 },
-	[MF_PLT_CLK2]		= { 0,		7,	-1 },
-	[9]					= { -1,		-1,	-1 },
-	[10]				= { -1,		-1,	-1 },
-	[11]				= { -1,		-1,	-1 },
-	[12]				= { -1,		-1,	-1 },
-	[13]				= { -1,		-1,	-1 },
-	[14]				= { -1,		-1,	-1 },
+	[MF_PLT_CLK0]		= { 0,		0,	-1, -1},
+	[PWM1]				= { 0,		1,	-1, -1},
+	[MF_PLT_CLK1]		= { 0,		2,	-1, -1},
+	[MF_PLT_CLK4]		= { 0,		3,	-1, -1},
+	[MF_PLT_CLK3]		= { 0,		4,	-1, -1},
+	[PWM0]				= { 0,		5,	-1, -1},
+	[MF_PLT_CLK5]		= { 0,		6,	-1, -1},
+	[MF_PLT_CLK2]		= { 0,		7,	-1, -1},
+	[9]					= { -1,		-1,	-1, -1},
+	[10]				= { -1,		-1,	-1, -1},
+	[11]				= { -1,		-1,	-1, -1},
+	[12]				= { -1,		-1,	-1, -1},
+	[13]				= { -1,		-1,	-1, -1},
+	[14]				= { -1,		-1,	-1, -1},
 
 	/*15~29*/
-	[SDMMC2_D3_CD_B]	= { 1,		0,	-1 },
-	[SDMMC1_CLK]		= { 1,		1,	-1 },
-	[SDMMC1_D0]			= { 1,		2,	-1 },
-	[SDMMC2_D1]			= { 1,		3,	-1 },
-	[SDMMC2_CLK]		= { 1,		4,	-1 },
-	[SDMMC1_D2]			= { 1,		5,	-1 },
-	[SDMMC2_D2]			= { 1,		6,	-1 },
-	[SDMMC2_CMD]		= { 1,		7,	-1 },
-	[SDMMC1_CMD]		= { 1,		8,	-1 },
-	[SDMMC1_D1]			= { 1,		9,	-1 },
-	[SDMMC2_D0]			= { 1,		10,	-1 },
-	[SDMMC1_D3_CD_B]	= { 1,		11,	-1 },
-	[27]				= { -1,		-1,	-1 },
-	[28]				= { -1,		-1,	-1 },
-	[29]				= { -1,		-1,	-1 },
+	[SDMMC2_D3_CD_B]	= { 1,		0,	-1, -1},
+	[SDMMC1_CLK]		= { 1,		1,	-1, -1},
+	[SDMMC1_D0]			= { 1,		2,	-1, -1},
+	[SDMMC2_D1]			= { 1,		3,	-1, -1},
+	[SDMMC2_CLK]		= { 1,		4,	-1, -1},
+	[SDMMC1_D2]			= { 1,		5,	-1, -1},
+	[SDMMC2_D2]			= { 1,		6,	-1, -1},
+	[SDMMC2_CMD]		= { 1,		7,	-1, -1},
+	[SDMMC1_CMD]		= { 1,		8,	-1, -1},
+	[SDMMC1_D1]			= { 1,		9,	-1, -1},
+	[SDMMC2_D0]			= { 1,		10,	-1, -1},
+	[SDMMC1_D3_CD_B]	= { 1,		11,	-1, -1},
+	[27]				= { -1,		-1,	-1, -1},
+	[28]				= { -1,		-1,	-1, -1},
+	[29]				= { -1,		-1,	-1, -1},
 
 	/*30~44*/
-	[SDMMC3_D1]			= { 2,		0,	-1 },
-	[SDMMC3_CLK]		= { 2,		1,	-1 },
-	[SDMMC3_D3]			= { 2,		2,	-1 },
-	[SDMMC3_D2]			= { 2,		3,	-1 },
-	[SDMMC3_CMD]		= { 2,		4,	-1 },
-	[SDMMC3_D0]			= { 2,		5,	-1 },
-	[36]				= { -1,		-1,	-1 },
-	[37]				= { -1,		-1,	-1 },
-	[38]				= { -1,		-1,	-1 },
-	[39]				= { -1,		-1,	-1 },
-	[40]				= { -1,		-1,	-1 },
-	[41]				= { -1,		-1,	-1 },
-	[42]				= { -1,		-1,	-1 },
-	[43]				= { -1,		-1,	-1 },
-	[44]				= { -1,		-1,	-1 },
+	[SDMMC3_D1]			= { 2,		0,	-1, -1},
+	[SDMMC3_CLK]		= { 2,		1,	-1, -1},
+	[SDMMC3_D3]			= { 2,		2,	-1, -1},
+	[SDMMC3_D2]			= { 2,		3,	-1, -1},
+	[SDMMC3_CMD]		= { 2,		4,	-1, -1},
+	[SDMMC3_D0]			= { 2,		5,	-1, -1},
+	[36]				= { -1,		-1,	-1, -1},
+	[37]				= { -1,		-1,	-1, -1},
+	[38]				= { -1,		-1,	-1, -1},
+	[39]				= { -1,		-1,	-1, -1},
+	[40]				= { -1,		-1,	-1, -1},
+	[41]				= { -1,		-1,	-1, -1},
+	[42]				= { -1,		-1,	-1, -1},
+	[43]				= { -1,		-1,	-1, -1},
+	[44]				= { -1,		-1,	-1, -1},
 
 	/*45~59*/
-	[MF_LPC_AD2]		= { 3,		0,	-1 },
-	[LPC_CLKRUNB]		= { 3,		1,	-1 },
-	[MF_LPC_AD0]		= { 3,		2,	-1 },
-	[LPC_FRAMEB]		= { 3,		3,	-1 },
-	[MF_LPC_CLKOUT1]	= { 3,		4,	-1 },
-	[MF_LPC_AD3]		= { 3,		5,	-1 },
-	[MF_LPC_CLKOUT0]	= { 3,		6,	-1 },
-	[MF_LPC_AD1]		= { 3,		7,	-1 },
-	[53]				= { -1,		-1,	-1 },
-	[54]				= { -1,		-1,	-1 },
-	[55]				= { -1,		-1,	-1 },
-	[56]				= { -1,		-1,	-1 },
-	[57]				= { -1,		-1,	-1 },
-	[58]				= { -1,		-1,	-1 },
-	[59]				= { -1,		-1,	-1 },
+	[MF_LPC_AD2]		= { 3,		0,	-1, -1},
+	[LPC_CLKRUNB]		= { 3,		1,	-1, -1},
+	[MF_LPC_AD0]		= { 3,		2,	-1, -1},
+	[LPC_FRAMEB]		= { 3,		3,	-1, -1},
+	[MF_LPC_CLKOUT1]	= { 3,		4,	-1, -1},
+	[MF_LPC_AD3]		= { 3,		5,	-1, -1},
+	[MF_LPC_CLKOUT0]	= { 3,		6,	-1, -1},
+	[MF_LPC_AD1]		= { 3,		7,	-1, -1},
+	[53]				= { -1,		-1,	-1, -1},
+	[54]				= { -1,		-1,	-1, -1},
+	[55]				= { -1,		-1,	-1, -1},
+	[56]				= { -1,		-1,	-1, -1},
+	[57]				= { -1,		-1,	-1, -1},
+	[58]				= { -1,		-1,	-1, -1},
+	[59]				= { -1,		-1,	-1, -1},
 
 	/*60~74*/
-	[SPI1_MISO]			= { 4,		0,	-1 },
-	[SPI1_CSO_B]		= { 4,		1,	-1 },
-	[SPI1_CLK]			= { 4,		2,	-1 },
-	[MMC1_D6]			= { 4,		3,	-1 },
-	[SPI1_MOSI]			= { 4,		4,	-1 },
-	[MMC1_D5]			= { 4,		5,	-1 },
-	[SPI1_CS1_B]		= { 4,		6,	-1 },
-	[MMC1_D4_SD_WE]		= { 4,		7,	-1 },
-	[MMC1_D7]			= { 4,		8,	-1 },
-	[MMC1_RCLK]			= { 4,		9,	-1 },
-	[70]				= { -1,		-1,	-1 },
-	[71]				= { -1,		-1,	-1 },
-	[72]				= { -1,		-1,	-1 },
-	[73]				= { -1,		-1,	-1 },
-	[74]				= { -1,		-1,	-1 },
+	[SPI1_MISO]			= { 4,		0,	-1, -1},
+	[SPI1_CSO_B]		= { 4,		1,	-1, -1},
+	[SPI1_CLK]			= { 4,		2,	-1, -1},
+	[MMC1_D6]			= { 4,		3,	-1, -1},
+	[SPI1_MOSI]			= { 4,		4,	-1, -1},
+	[MMC1_D5]			= { 4,		5,	-1, -1},
+	[SPI1_CS1_B]		= { 4,		6,	-1, -1},
+	[MMC1_D4_SD_WE]		= { 4,		7,	-1, -1},
+	[MMC1_D7]			= { 4,		8,	-1, -1},
+	[MMC1_RCLK]			= { 4,		9,	-1, -1},
+	[70]				= { -1,		-1,	-1, -1},
+	[71]				= { -1,		-1,	-1, -1},
+	[72]				= { -1,		-1,	-1, -1},
+	[73]				= { -1,		-1,	-1, -1},
+	[74]				= { -1,		-1,	-1, -1},
 
 	/*75~85*/
-	[USB_OC1_B]			= { 5,		0,	-1 },
-	[PMU_RESETBUTTON_B]	= { 5,		1,	-1 },
-	[GPIO_ALERT]		= { 5,		2,	-1 },
-	[SDMMC3_PWR_EN_B]	= { 5,		3,	-1 },
-	[ILB_SERIRQ]		= { 5,		4,	-1 },
-	[USB_OC0_B]			= { 5,		5,	-1 },
-	[SDMMC3_CD_B]		= { 5,		6,	-1 },
-	[SPKR]				= { 5,		7,	-1 },
-	[SUSPWRDNACK]		= { 5,		8,	-1 },
-	[SPARE_PIN]			= { 5,		9,	-1 },
-	[SDMMC3_1P8_EN]		= { 5,		10,	-1 },
+	[USB_OC1_B]			= { 5,		0,	-1, -1},
+	[PMU_RESETBUTTON_B]	= { 5,		1,	-1, -1},
+	[GPIO_ALERT]		= { 5,		2,	-1, -1},
+	[SDMMC3_PWR_EN_B]	= { 5,		3,	-1, -1},
+	[ILB_SERIRQ]		= { 5,		4,	-1, -1},
+	[USB_OC0_B]			= { 5,		5,	-1, -1},
+	[SDMMC3_CD_B]		= { 5,		6,	-1, -1},
+	[SPKR]				= { 5,		7,	-1, -1},
+	[SUSPWRDNACK]		= { 5,		8,	-1, -1},
+	[SPARE_PIN]			= { 5,		9,	-1, -1},
+	[SDMMC3_1P8_EN]		= { 5,		10,	-1, -1},
 };
 
 static struct gpio_pad_info east_pads_info[] = {
 	/*0~14*/
-	[PMU_SLP_S3_B]		= { 0,		0,	-1 },
-	[PMU_BATLOW_B]		= { 0,		1,	-1 },
-	[SUS_STAT_B]		= { 0,		2,	-1 },
-	[PMU_SLP_S0IX_B]	= { 0,		3,	-1 },
-	[PMU_AC_PRESENT]	= { 0,		4,	-1 },
-	[PMU_PLTRST_B]		= { 0,		5,	-1 },
-	[PMU_SUSCLK]		= { 0,		6,	-1 },
-	[PMU_SLP_LAN_B]		= { 0,		7,	-1 },
-	[PMU_PWRBTN_B]		= { 0,		8,	-1 },
-	[PMU_SLP_S4_B]		= { 0,		9,	-1 },
-	[PMU_WAKE_B]		= { 0,		10,	-1 },
-	[PMU_WAKE_LAN_B]	= { 0,		11,	-1 },
-	[12]				= { -1,		-1,	-1 },
-	[13]				= { -1,		-1,	-1 },
-	[14]				= { -1,		-1,	-1 },
+	[PMU_SLP_S3_B]		= { 0,		0,	-1, -1},
+	[PMU_BATLOW_B]		= { 0,		1,	-1, -1},
+	[SUS_STAT_B]		= { 0,		2,	-1, -1},
+	[PMU_SLP_S0IX_B]	= { 0,		3,	-1, -1},
+	[PMU_AC_PRESENT]	= { 0,		4,	-1, -1},
+	[PMU_PLTRST_B]		= { 0,		5,	-1, -1},
+	[PMU_SUSCLK]		= { 0,		6,	-1, -1},
+	[PMU_SLP_LAN_B]		= { 0,		7,	-1, -1},
+	[PMU_PWRBTN_B]		= { 0,		8,	-1, -1},
+	[PMU_SLP_S4_B]		= { 0,		9,	-1, -1},
+	[PMU_WAKE_B]		= { 0,		10,	-1, -1},
+	[PMU_WAKE_LAN_B]	= { 0,		11,	-1, -1},
+	[12]				= { -1,		-1,	-1, -1},
+	[13]				= { -1,		-1,	-1, -1},
+	[14]				= { -1,		-1,	-1, -1},
 
 	/*15~26*/
-	[MF_ISH_GPIO_3]		= { 1,		0,	-1 },
-	[MF_ISH_GPIO_7]		= { 1,		1,	-1 },
-	[MF_ISH_I2C1_SCL]	= { 1,		2,	-1 },
-	[MF_ISH_GPIO_1]		= { 1,		3,	-1 },
-	[MF_ISH_GPIO_5]		= { 1,		4,	-1 },
-	[MF_ISH_GPIO_9]		= { 1,		5,	-1 },
-	[MF_ISH_GPIO_0]		= { 1,		6,	-1 },
-	[MF_ISH_GPIO_4]		= { 1,		7,	-1 },
-	[MF_ISH_GPIO_8]		= { 1,		8,	-1 },
-	[MF_ISH_GPIO_2]		= { 1,		9,	-1 },
-	[MF_ISH_GPIO_6]		= { 1,		10,	-1 },
-	[MF_ISH_I2C1_SDA]	= { 1,		11,	-1 },
+	[MF_ISH_GPIO_3]		= { 1,		0,	-1, -1},
+	[MF_ISH_GPIO_7]		= { 1,		1,	-1, -1},
+	[MF_ISH_I2C1_SCL]	= { 1,		2,	-1, -1},
+	[MF_ISH_GPIO_1]		= { 1,		3,	-1, -1},
+	[MF_ISH_GPIO_5]		= { 1,		4,	-1, -1},
+	[MF_ISH_GPIO_9]		= { 1,		5,	-1, -1},
+	[MF_ISH_GPIO_0]		= { 1,		6,	-1, -1},
+	[MF_ISH_GPIO_4]		= { 1,		7,	-1, -1},
+	[MF_ISH_GPIO_8]		= { 1,		8,	-1, -1},
+	[MF_ISH_GPIO_2]		= { 1,		9,	-1, -1},
+	[MF_ISH_GPIO_6]		= { 1,		10,	-1, -1},
+	[MF_ISH_I2C1_SDA]	= { 1,		11,	-1, -1},
 };
 
 static struct gpio_pad_info southwest_pads_info[] = {
 	/*0~14*/
-	[FST_SPI_D2]		= { 0,		0,	-1 },
-	[FST_SPI_D0]		= { 0,		1,	-1 },
-	[FST_SPI_CLK]		= { 0,		2,	-1 },
-	[FST_SPI_D3]		= { 0,		3,	-1 },
-	[FST_SPI_CS1_B]		= { 0,		4,	-1 },
-	[FST_SPI_D1]		= { 0,		5,	-1 },
-	[FST_SPI_CS0_B]		= { 0,		6,	-1 },
-	[FST_SPI_CS2_B]		= { 0,		7,	-1 },
-	[8]					= { -1,		-1,	-1 },
-	[9]					= { -1,		-1,	-1 },
-	[10]				= { -1,		-1,	-1 },
-	[11]				= { -1,		-1,	-1 },
-	[12]				= { -1,		-1,	-1 },
-	[13]				= { -1,		-1,	-1 },
-	[14]				= { -1,		-1,	-1 },
+	[FST_SPI_D2]		= { 0,		0,	-1, -1},
+	[FST_SPI_D0]		= { 0,		1,	-1, -1},
+	[FST_SPI_CLK]		= { 0,		2,	-1, -1},
+	[FST_SPI_D3]		= { 0,		3,	-1, -1},
+	[FST_SPI_CS1_B]		= { 0,		4,	-1, -1},
+	[FST_SPI_D1]		= { 0,		5,	-1, -1},
+	[FST_SPI_CS0_B]		= { 0,		6,	-1, -1},
+	[FST_SPI_CS2_B]		= { 0,		7,	-1, -1},
+	[8]					= { -1,		-1,	-1, -1},
+	[9]					= { -1,		-1,	-1, -1},
+	[10]				= { -1,		-1,	-1, -1},
+	[11]				= { -1,		-1,	-1, -1},
+	[12]				= { -1,		-1,	-1, -1},
+	[13]				= { -1,		-1,	-1, -1},
+	[14]				= { -1,		-1,	-1, -1},
 
 	/*15~29*/
-	[UART1_RTS_B]		= { 1,		0,	-1 },
-	[UART1_RXD]			= { 1,		1,	-1 },
-	[UART2_RXD]			= { 1,		2,	-1 },
-	[UART1_CTS_B]		= { 1,		3,	-1 },
-	[UART2_RTS_B]		= { 1,		4,	-1 },
-	[UART1_TXD]			= { 1,		5,	-1 },
-	[UART2_TXD]			= { 1,		6,	-1 },
-	[UART2_CTS_B]		= { 1,		7,	-1 },
-	[23]				= { -1,		-1,	-1 },
-	[24]				= { -1,		-1,	-1 },
-	[25]				= { -1,		-1,	-1 },
-	[26]				= { -1,		-1,	-1 },
-	[27]				= { -1,		-1,	-1 },
-	[28]				= { -1,		-1,	-1 },
-	[29]				= { -1,		-1,	-1 },
+	[UART1_RTS_B]		= { 1,		0,	-1, 1},
+	[UART1_RXD]			= { 1,		1,	-1, 1},
+	[UART2_RXD]			= { 1,		2,	-1, 1},
+	[UART1_CTS_B]		= { 1,		3,	-1, -1},
+	[UART2_RTS_B]		= { 1,		4,	-1, 1},
+	[UART1_TXD]			= { 1,		5,	-1, -1},
+	[UART2_TXD]			= { 1,		6,	-1, -1},
+	[UART2_CTS_B]		= { 1,		7,	-1, -1},
+	[23]				= { -1,		-1,	-1, -1},
+	[24]				= { -1,		-1,	-1, -1},
+	[25]				= { -1,		-1,	-1, -1},
+	[26]				= { -1,		-1,	-1, -1},
+	[27]				= { -1,		-1,	-1, -1},
+	[28]				= { -1,		-1,	-1, -1},
+	[29]				= { -1,		-1,	-1, -1},
 
 	/*30~44*/
-	[MF_HDA_CLK]		= { 2,		0,	-1 },
-	[MF_HDA_RSTB]		= { 2,		1,	-1 },
-	[MF_HDA_SDIO]		= { 2,		2,	-1 },
-	[MF_HDA_SDO]		= { 2,		3,	-1 },
-	[MF_HDA_DOCKRSTB]	= { 2,		4,	-1 },
-	[MF_HDA_SYNC]		= { 2,		5,	-1 },
-	[MF_HDA_SDI1]		= { 2,		6,	-1 },
-	[MF_HDA_DOCKENB]	= { 2,		7,	-1 },
-	[38]				= { -1,		-1,	-1 },
-	[39]				= { -1,		-1,	-1 },
-	[40]				= { -1,		-1,	-1 },
-	[41]				= { -1,		-1,	-1 },
-	[42]				= { -1,		-1,	-1 },
-	[43]				= { -1,		-1,	-1 },
-	[44]				= { -1,		-1,	-1 },
+	[MF_HDA_CLK]		= { 2,		0,	-1, -1},
+	[MF_HDA_RSTB]		= { 2,		1,	-1, -1},
+	[MF_HDA_SDIO]		= { 2,		2,	-1, -1},
+	[MF_HDA_SDO]		= { 2,		3,	-1, -1},
+	[MF_HDA_DOCKRSTB]	= { 2,		4,	-1, -1},
+	[MF_HDA_SYNC]		= { 2,		5,	-1, -1},
+	[MF_HDA_SDI1]		= { 2,		6,	-1, -1},
+	[MF_HDA_DOCKENB]	= { 2,		7,	-1, -1},
+	[38]				= { -1,		-1,	-1, -1},
+	[39]				= { -1,		-1,	-1, -1},
+	[40]				= { -1,		-1,	-1, -1},
+	[41]				= { -1,		-1,	-1, -1},
+	[42]				= { -1,		-1,	-1, -1},
+	[43]				= { -1,		-1,	-1, -1},
+	[44]				= { -1,		-1,	-1, -1},
 
 	/*45~59*/
-	[I2C5_SDA]			= { 3,		0,	-1 },
-	[I2C4_SDA]			= { 3,		1,	-1 },
-	[I2C6_SDA]			= { 3,		2,	-1 },
-	[I2C5_SCL]			= { 3,		3,	-1 },
-	[I2C_NFC_SDA]		= { 3,		4,	-1 },
-	[I2C4_SCL]			= { 3,		5,	-1 },
-	[I2C6_SCL]			= { 3,		6,	-1 },
-	[I2C_NFC_SCL]		= { 3,		7,	-1 },
-	[53]				= { -1,		-1,	-1 },
-	[54]				= { -1,		-1,	-1 },
-	[55]				= { -1,		-1,	-1 },
-	[56]				= { -1,		-1,	-1 },
-	[57]				= { -1,		-1,	-1 },
-	[58]				= { -1,		-1,	-1 },
-	[59]				= { -1,		-1,	-1 },
+	[I2C5_SDA]			= { 3,		0,	-1, -1},
+	[I2C4_SDA]			= { 3,		1,	-1, -1},
+	[I2C6_SDA]			= { 3,		2,	-1, -1},
+	[I2C5_SCL]			= { 3,		3,	-1, -1},
+	[I2C_NFC_SDA]		= { 3,		4,	-1, -1},
+	[I2C4_SCL]			= { 3,		5,	-1, -1},
+	[I2C6_SCL]			= { 3,		6,	-1, -1},
+	[I2C_NFC_SCL]		= { 3,		7,	-1, -1},
+	[53]				= { -1,		-1,	-1, -1},
+	[54]				= { -1,		-1,	-1, -1},
+	[55]				= { -1,		-1,	-1, -1},
+	[56]				= { -1,		-1,	-1, -1},
+	[57]				= { -1,		-1,	-1, -1},
+	[58]				= { -1,		-1,	-1, -1},
+	[59]				= { -1,		-1,	-1, -1},
 
 	/*60~74*/
-	[I2C1_SDA]			= { 4,		0,	-1 },
-	[I2C0_SDA]			= { 4,		1,	-1 },
-	[I2C2_SDA]			= { 4,		2,	-1 },
-	[I2C1_SCL]			= { 4,		3,	-1 },
-	[I2C3_SDA]			= { 4,		4,	-1 },
-	[I2C0_SCL]			= { 4,		5,	-1 },
-	[I2C2_SCL]			= { 4,		6,	-1 },
-	[I2C3_SCL]			= { 4,		7,	-1 },
-	[68]				= { -1,		-1,	-1 },
-	[69]				= { -1,		-1,	-1 },
-	[70]				= { -1,		-1,	-1 },
-	[71]				= { -1,		-1,	-1 },
-	[72]				= { -1,		-1,	-1 },
-	[73]				= { -1,		-1,	-1 },
-	[74]				= { -1,		-1,	-1 },
+	[I2C1_SDA]			= { 4,		0,	-1, -1},
+	[I2C0_SDA]			= { 4,		1,	-1, -1},
+	[I2C2_SDA]			= { 4,		2,	-1, -1},
+	[I2C1_SCL]			= { 4,		3,	-1, -1},
+	[I2C3_SDA]			= { 4,		4,	-1, -1},
+	[I2C0_SCL]			= { 4,		5,	-1, -1},
+	[I2C2_SCL]			= { 4,		6,	-1, -1},
+	[I2C3_SCL]			= { 4,		7,	-1, -1},
+	[68]				= { -1,		-1,	-1, -1},
+	[69]				= { -1,		-1,	-1, -1},
+	[70]				= { -1,		-1,	-1, -1},
+	[71]				= { -1,		-1,	-1, -1},
+	[72]				= { -1,		-1,	-1, -1},
+	[73]				= { -1,		-1,	-1, -1},
+	[74]				= { -1,		-1,	-1, -1},
 
 	/*75~89*/
-	[SATA_GP0]			= { 5,		0,	-1 },
-	[SATA_GP1]			= { 5,		1,	-1 },
-	[SATA_LEDN]			= { 5,		2,	-1 },
-	[SATA_GP2]			= { 5,		3,	-1 },
-	[MF_SMB_ALERTB]		= { 5,		4,	-1 },
-	[SATA_GP3]			= { 5,		5,	-1 },
-	[MF_SMB_CLK]		= { 5,		6,	-1 },
-	[MF_SMB_DATA]		= { 5,		7,	-1 },
-	[83]				= { -1,		-1,	-1 },
-	[84]				= { -1,		-1,	-1 },
-	[85]				= { -1,		-1,	-1 },
-	[86]				= { -1,		-1,	-1 },
-	[87]				= { -1,		-1,	-1 },
-	[88]				= { -1,		-1,	-1 },
-	[89]				= { -1,		-1,	-1 },
+	[SATA_GP0]			= { 5,		0,	-1, -1},
+	[SATA_GP1]			= { 5,		1,	-1, -1},
+	[SATA_LEDN]			= { 5,		2,	-1, -1},
+	[SATA_GP2]			= { 5,		3,	-1, -1},
+	[MF_SMB_ALERTB]		= { 5,		4,	-1, -1},
+	[SATA_GP3]			= { 5,		5,	-1, -1},
+	[MF_SMB_CLK]		= { 5,		6,	-1, -1},
+	[MF_SMB_DATA]		= { 5,		7,	-1, -1},
+	[83]				= { -1,		-1,	-1, -1},
+	[84]				= { -1,		-1,	-1, -1},
+	[85]				= { -1,		-1,	-1, -1},
+	[86]				= { -1,		-1,	-1, -1},
+	[87]				= { -1,		-1,	-1, -1},
+	[88]				= { -1,		-1,	-1, -1},
+	[89]				= { -1,		-1,	-1, -1},
 
 	/*90~97*/
-	[PCIE_CLKREQ0B]		= { 6,		0,	-1 },
-	[PCIE_CLKREQ1B]		= { 6,		1,	-1 },
-	[GP_SSP_2_CLK]		= { 6,		2,	-1 },
-	[PCIE_CLKREQ2B]		= { 6,		3,	-1 },
-	[GP_SSP_2_RXD]		= { 6,		4,	-1 },
-	[PCIE_CLKREQ3B]		= { 6,		5,	-1 },
-	[GP_SSP_2_FS]		= { 6,		6,	-1 },
-	[GP_SSP_2_TXD]		= { 6,		7,	-1 },
+	[PCIE_CLKREQ0B]		= { 6,		0,	-1, -1},
+	[PCIE_CLKREQ1B]		= { 6,		1,	-1, -1},
+	[GP_SSP_2_CLK]		= { 6,		2,	-1, -1},
+	[PCIE_CLKREQ2B]		= { 6,		3,	-1, -1},
+	[GP_SSP_2_RXD]		= { 6,		4,	-1, -1},
+	[PCIE_CLKREQ3B]		= { 6,		5,	-1, -1},
+	[GP_SSP_2_FS]		= { 6,		6,	-1, -1},
+	[GP_SSP_2_TXD]		= { 6,		7,	-1, -1},
 };
 
 static struct gpio_pad_info virtual_pads_info[] = {
-	[VIRTUAL0]		= { 0,		0,	-1 },
-	[VIRTUAL1]		= { 0,		1,	-1 },
-	[VIRTUAL2]		= { 0,		2,	-1 },
-	[VIRTUAL3]		= { 0,		3,	-1 },
-	[VIRTUAL4]		= { 0,		4,	-1 },
-	[VIRTUAL5]		= { 0,		5,	-1 },
-	[VIRTUAL6]		= { 0,		6,	-1 },
-	[VIRTUAL7]		= { 0,		7,	-1 },
+	[VIRTUAL0]		= { 0,		0,	-1, -1},
+	[VIRTUAL1]		= { 0,		1,	-1, -1},
+	[VIRTUAL2]		= { 0,		2,	-1, -1},
+	[VIRTUAL3]		= { 0,		3,	-1, -1},
+	[VIRTUAL4]		= { 0,		4,	-1, -1},
+	[VIRTUAL5]		= { 0,		5,	-1, -1},
+	[VIRTUAL6]		= { 0,		6,	-1, -1},
+	[VIRTUAL7]		= { 0,		7,	-1, -1},
 };
 
 static struct gpio_bank_pnp chv_banks_pnp[] = {
@@ -775,9 +780,13 @@ static int chv_gpio_request(struct gpio_chip *chip, unsigned offset)
 
 	dev_dbg(&cg->pdev->dev, "chv_gpio_request %d\n", offset);
 
-	/*change it to GPIO mode*/
-	value = chv_readl(reg) | CV_GPIO_EN;
-	chv_writel(value, reg);
+	if (cg->pad_info[offset].switch_en > 0) {
+		/*change it to GPIO mode*/
+		value = chv_readl(reg) | CV_GPIO_EN;
+		chv_writel(value, reg);
+		dev_info(&cg->pdev->dev,
+			"pin %d force set to gpio mode\n", offset);
+	}
 
 	return 0;
 }
@@ -793,9 +802,13 @@ static void chv_gpio_free(struct gpio_chip *chip, unsigned offset)
 
 	dev_dbg(&cg->pdev->dev, "chv_gpio_free %d\n", offset);
 
-	/*clear GPIOEn bit to let Pad mode take effect*/
-	value = chv_readl(reg) & (~CV_GPIO_EN);
-	chv_writel(value, reg);
+	if (cg->pad_info[offset].switch_en > 0) {
+		/*clear GPIOEn bit to let Pad mode take effect*/
+		value = chv_readl(reg) & (~CV_GPIO_EN);
+		chv_writel(value, reg);
+		dev_info(&cg->pdev->dev,
+			"pin %d switch back to native mode\n", offset);
+	}
 }
 
 static void chv_update_irq_type(struct chv_gpio *cg, unsigned type,
-- 
2.1.0

